### PR TITLE
Distribute weight to children when closing stream

### PR DESCRIPTION
**Pull request by kapsterio**
Motivation:

As suggested in [section 5.3.4 in http2 spec](https://datatracker.ietf.org/doc/html/rfc7540#section-5.3.4):

> When a stream is removed from the dependency tree, its dependencies can be moved to become dependent on the parent of the closed stream. The weights of new dependencies are recalculated by distributing the weight of the dependency of the closed stream proportionally based on the weights of its dependencies.

For example, we have stream A and B depend on connection stream with default weights (16), and stream C depends on A with max weight (256). When stream A was closed, we move stream C to become dependent on the connection stream, then we should distribute the weight of stream A to its children (only stream C), so the new weight of stream C will be 16. If we keep the weight of stream C unchanged, it will get more resource than stream B

Modification:
- distribute weight to its children when closing a stream
- add a unit test for the case above and fix other related unit tests

Result:

More spec-compliant and more appropriate stream reprioritization

### CODE REVIEW

File: codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java [hunk 1]
```diff
@@ -553,13 +554,24 @@ void removeChild(State child) {
                 // Move up any grand children to be directly dependent on this node.
                 Iterator<IntObjectMap.PrimitiveEntry<State>> itr = child.children.entries().iterator();
                 while (itr.hasNext()) {
-                    takeChild(itr, itr.next().value(), false, events);
+                    //distribute the weight of child to it's dependency base on the proportion
+                    State dependency = itr.next().value();
+                    dependency.weight = (short) ((double) dependency.weight / totalWeight * child.weight);
```
Comment (NiteshKant) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java:559]: This is the first time I am reading this code, so unsure whether this is actually meaningful but does changing weight has to also consider `totalQueuedWeights` of the parent?

Comment (ejona86) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java:559]: I don't think so. `totalQueuedWeights` just contains a stream's child weights.

[the remark this answers is not in the document]
Comment (kapsterio) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java]: Done, PTAL

Comment (kapsterio) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java]: To be more specific, `totalQueuedWeights` only contains the total weights of children in a stream's `pseudoTimeQueue`, if the `activeCountForTree` of a child is 0, it's weight will not been counted into `totalQueuedWeights`.

**Review by ejona86 — APPROVED** [review]

**Comment by ejona86** [comment]
Not merging yet just to give a bit more time for review. It's been a while since we were in this code, so I'd be easy to miss something (`totalQueuedWeights` gave me a scare). That said, `updateDependencyTree()` seems to be a good reference of what is necessary when changing weights, and nothing jumped out as missing.

**Comment by NiteshKant** [comment]
@ejona86 , @kapsterio based on my limited knowledge of this code, this looks OK but yeah I would agree that lets wait for a bit for others to have a look. @Scottmitch may have inputs.

**Comment by kapsterio** [comment]
Sure, lets wait, and thanks you guys for quick response

File: codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java [hunk 2]
```diff
@@ -553,13 +554,24 @@ void removeChild(State child) {
                 // Move up any grand children to be directly dependent on this node.
                 Iterator<IntObjectMap.PrimitiveEntry<State>> itr = child.children.entries().iterator();
                 while (itr.hasNext()) {
-                    takeChild(itr, itr.next().value(), false, events);
+                    //distribute the weight of child to it's dependency base on the proportion
```
Comment (trustin) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java:557]:
```suggestion
                    // Redistribute the weight of child to its dependency proportionally.
```

File: codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java [hunk 3]
```diff
@@ -545,6 +545,7 @@ void takeChild(Iterator<IntObjectMap.PrimitiveEntry<State>> childItr, State chil
          * Removes the child priority and moves any of its dependencies to being direct dependencies on this node.
          */
         void removeChild(State child) {
+            long totalWeight = child.getTotalWeight();
```
Comment (trustin) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java:548]: Could calculate the total weight only when necessary. How about moving this right before the `while` block below? We could also skip the entire loop and calculation if `child.children` is empty.

File: codec-http2/src/test/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributorDependencyTreeTest.java [hunk 4]
```diff
@@ -666,6 +668,7 @@ public void removeWithPrioritizableDependentsShouldNotRestructureTree() throws E
         setPriority(streamD.id(), streamB.id(), DEFAULT_PRIORITY_WEIGHT, false);
 
         // Default removal policy will cause it to be removed immediately.
+        // close stream B will distribute it's weight to children (C & D) equally
```
Comment (trustin) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributorDependencyTreeTest.java:671]:
```suggestion
        // Closing streamB will distribute its weight to the children (C & D) equally.
```

**Review by normanmaurer — APPROVED** [review]

File: codec-http2/src/test/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributorDependencyTreeTest.java [hunk 5]
```diff
@@ -700,6 +704,7 @@ public void closeWithNoPrioritizableDependentsShouldRestructureTree() throws Exc
 
         // Close internal nodes, leave 1 leaf node open, the only remaining stream is the one that is not closed (E).
         streamA.close();
+        //close stream B will distribute it's weight to children (C & D) equally
```
Comment (trustin) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributorDependencyTreeTest.java:707]:
```suggestion
        // Closing streamB will distribute its weight to the children (C & D) equally.
```

File: codec-http2/src/test/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributorDependencyTreeTest.java [hunk 6]
```diff
@@ -730,6 +756,7 @@ public void priorityChangeWithNoPrioritizableDependentsShouldRestructureTree() t
 
         // Leave leaf nodes open (E & F)
         streamA.close();
+        //close stream B will distribute it's weight to children (C & D) equally
```
Comment (trustin) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributorDependencyTreeTest.java:759]:
```suggestion
        // Closing streamB will distribute its weight to the children (C & D) equally.
```

File: codec-http2/src/test/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributorDependencyTreeTest.java [hunk 7]
```diff
@@ -709,10 +714,31 @@ public void closeWithNoPrioritizableDependentsShouldRestructureTree() throws Exc
...
 
+    @Test
+    public void closeStreamWithChildrenShouldRedistributeWeightToChildren() throws Exception {
+        Http2Stream streamA = connection.local().createStream(1, false);
+        Http2Stream streamB = connection.local().createStream(3, false);
+        Http2Stream streamC = connection.local().createStream(5, false);
+        Http2Stream streamD = connection.local().createStream(7, false);
+
+        setPriority(streamC.id(), streamA.id(), MAX_WEIGHT, false);
+        setPriority(streamD.id(), streamA.id(), MAX_WEIGHT, false);
```
Comment (trustin) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributorDependencyTreeTest.java:730]: How about making this slightly more complex, by making each parent stream have 3 children?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java (hunk 3 above)
Comment (kapsterio) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java:548]: yeah, and thanks @trustin  for your advices, I'll make a change about it

**Comment by kapsterio** [comment]
@trustin PTAL

File: codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java [hunk 8]
```diff
@@ -550,16 +550,30 @@ void removeChild(State child) {
...
 
-                // Move up any grand children to be directly dependent on this node.
-                Iterator<IntObjectMap.PrimitiveEntry<State>> itr = child.children.entries().iterator();
-                while (itr.hasNext()) {
-                    takeChild(itr, itr.next().value(), false, events);
+                if (!child.children.isEmpty()) {
+                    // Move up any grand children to be directly dependent on this node.
+                    Iterator<IntObjectMap.PrimitiveEntry<State>> itr = child.children.entries().iterator();
+                    long totalWeight = child.getTotalWeight();
+                    while (itr.hasNext()) {
```
Comment (trustin) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java:557]: could use `do ... while (itr.hasNext());`

**Review by trustin — APPROVED** [review]
Just one nit. Thanks!

**Comment by normanmaurer** [comment]
@kapsterio please fix the check style.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java (hunk 8 above)
Comment (kapsterio) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java:557]: yes, sure

**Comment by kapsterio** [comment]
Done @trustin  @normanmaurer

File: codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java [hunk 9]
```diff
@@ -550,16 +550,30 @@ void removeChild(State child) {
...
+                        dependency.weight = (short) ((int) dependency.weight * child.weight / totalWeight);
+                        takeChild(itr, dependency, false, events);
+                    } while (itr.hasNext());
                 }
 
                 notifyParentChanged(events);
             }
         }
 
+        long getTotalWeight() {
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java:569]: nit: make this private?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java [hunk 10]
```diff
@@ -550,16 +550,30 @@ void removeChild(State child) {
...
-                while (itr.hasNext()) {
-                    takeChild(itr, itr.next().value(), false, events);
+                if (!child.children.isEmpty()) {
+                    // Move up any grand children to be directly dependent on this node.
+                    Iterator<IntObjectMap.PrimitiveEntry<State>> itr = child.children.entries().iterator();
+                    long totalWeight = child.getTotalWeight();
+                    do {
+                        // Redistribute the weight of child to its dependency proportionally.
+                        State dependency = itr.next().value();
+                        dependency.weight = (short) ((int) dependency.weight * child.weight / totalWeight);
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java:560]: Ensure the result isn't less than 1 (e.g. `Math.max(1, ...)`)

https://datatracker.ietf.org/doc/html/rfc7540#section-5.3.2
>  All dependent streams are allocated an integer weight between 1 and
   256 (inclusive).

**Review by Scottmitch — APPROVED** [review]

File: codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java (hunk 10 above)
Comment (kapsterio) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java]: sure, this is necessary

File: codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java (hunk 9 above)
Comment (kapsterio) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java]: done

**Comment by normanmaurer** [comment]
@kapsterio this looks good now... Can you please sign our ICLA and let me know once done:

https://netty.io/s/icla

**Comment by kapsterio** [comment]
@normanmaurer Cool, already signed it before

**Comment by normanmaurer** [comment]
@kapsterio thanks a lot !

**Pull request merged**